core/src/main/scala/org/apache/spark/util/UninterruptibleThread.scala

```diff
@@ -69,10 +75,22 @@ private[spark] class UninterruptibleThread(
     }
 
     uninterruptibleLock.synchronized {
+      uninterruptible = true
+    }
+
+    while (uninterruptibleLock.synchronized {
       // Clear the interrupted status if it's set.
       shouldInterruptThread = Thread.interrupted() || shouldInterruptThread
-      uninterruptible = true
+      // wait for super.interrupt() to be called
+      !shouldInterruptThread && awaitInterruptThread }) {
+      try {
+        Thread.sleep(100)
```

Comment:
Previously, we could move forward as soon as the lock is released, but now we may wait up to 100 milliseconds. I think this is a regression. It's also more complicated as we need to handle `InterruptedException` from this sleep. Can you reconsider my CountDownLatch proposal?

Reply:
We can't move forward until `super.interrupt()` is called. 

I don't see any difference between awaiting on `CountDownLatch` and `Thread.sleep()`. Both would block till `super.interrupt()` is called and in your solution `await()` would exit on `InterruptedException`, not because `countDown()` was called and may be blocked for 100 milliseconds or longer in the similar way.

It is also quite unlikely that sleep would block for 100 milliseconds unless the thread that calls `super.interrupt()` is starved.

Reply:
A rough sketch would be something like this (Ignore impact on comments, etc - this is illustrative).
https://github.com/apache/spark/compare/master...mridulm:spark:mridulm/vrozov_uninterruptible-option

Please take a look @vrozov - given you have a lot of context into this.

Reply:
@mridulm 

>The concern is not just that sleep is called, but the cost of sleep - as opposed to using some sort of conditioned wait.

I don't see a big difference between interrupting sleep or waiting on a condition or any other thread synchronization. In case of `sleep()` we don't expect it to finish normally and in the most common case `sleep()` won't be even called and if it is called, it will exit on `InterruptedException` almost immediately.

Reply:
`sleep` has much higher overhead than parking a thread - not to mention it will suspend for atleast the specified duration, while unpark can come really quickly.

Reply:
+CC @Ngone51 - since you finished review - thoughts on the option above which does not need a sleep ?

Reply:
> sleep has a slightly higher overhead than parking/unparking a thread.

I am not convinced that `sleep()` has higher overhead and that it is important in this case. `sleep()` in this case serves two goals: wait for the interrupt and clear interrupt. It also simplifies logic for `UninterruptibleThread.interrupt()`.

Reply:
Given that a thead instance can be reused with a thread pool, I feel the current bool flag + sleep is safer than a signal/condition object with await.

Reply:
Not sure why that would impact either choices - both sleep and await are entered only when there is lock contention

Reply:
As I said both `sleep()` and `await()` are functionally equivalent. At the same time using `synchronized` on a `uninterruptibleLock` is a better option here as the compiler provides guarantee that the monitor is released when the control goes out of a synchronized block. With `ReentrantReadWriteLock` it is necessary to add code that provides such guarantee for no added functionality.